src/MainWindow.vala

```diff
@@ -370,6 +356,83 @@ public class Tasks.MainWindow : Hdy.ApplicationWindow {
         });
     }
 
+    private void add_collection_of_source (E.Source source, E.SourceRegistry registry) {
+        var collection_source = registry.find_extension (source, E.SOURCE_EXTENSION_COLLECTION);
+
+        if (collection_source == null && source.has_extension (E.SOURCE_EXTENSION_TASK_LIST)) {
+            var task_list_extension = (E.SourceTaskList) source.get_extension (E.SOURCE_EXTENSION_TASK_LIST);
+
+            if ("local" == task_list_extension.backend_name) {
+                try {
+                    collection_source = new E.Source.with_uid ("local", null);
+                    var collection_source_tasklist_extension = (E.SourceTaskList) collection_source.get_extension (E.SOURCE_EXTENSION_TASK_LIST);
+                    collection_source_tasklist_extension.backend_name = "local";
+                    collection_source.parent = "local-stub";
+                    collection_source.display_name = get_collection_source_display_name (source);
+                } catch (Error e) {
+                    warning (e.message);
+                }
+            }
+        }
+
+        if (collection_source == null) {
+            return;
+        }
+
+        lock (collection_sources) {
+            if (collection_sources == null) {
+                collection_sources = new Gee.HashSet<E.Source> (Util.esource_hash_func, Util.esource_equal_func);
+            }
+            E.SourceTaskList collection_source_tasklist_extension = (E.SourceTaskList)collection_source.get_extension (E.SOURCE_EXTENSION_TASK_LIST);
+
+            if (collection_sources.contains (collection_source) || !collection_source.enabled || !collection_source_tasklist_extension.selected) {
+                return;
+            }
+            collection_sources.add (collection_source);
+        }
+        update_add_tasklist_buttonbox ();
+    }
+
+    private void update_add_tasklist_buttonbox () {
+        foreach (var child in add_tasklist_buttonbox.get_children ()) {
+            add_tasklist_buttonbox.remove (child);
+        }
```

Comment:
This seems really inefficient to remove all the items and then add them back. It would probably be better to have separate add and remove functions if you're worried about making sure this list doesn't contain things that have been removed